Sources/Jobs/Scheduler/RecurrenceRule/RecurrenceRule.swift

```diff
@@ -0,0 +1,413 @@
+        try yearConstraint.addToSet(years)
+        return self
+    }
+
+    // ranges
+    public func whenSecondInRange(lowerBound: Int, upperBound: Int) throws -> Self {
+        try secondConstraint.createRange(lowerBound: lowerBound, upperBound: upperBound)
+
+        return self
+    }
+
+    public func whenMinuteInRange(lowerBound: Int, upperBound: Int) throws -> Self {
+        try minuteConstraint.createRange(lowerBound: lowerBound, upperBound: upperBound)
+
+        return self
+    }
+
+    public func whenHourInRange(lowerBound: Int, upperBound: Int) throws -> Self {
+        try hourConstraint.createRange(lowerBound: lowerBound, upperBound: upperBound)
+
+        return self
+    }
+
+    public func whenDayOfWeekInRange(lowerBound: Int, upperBound: Int) throws -> Self {
+        try dayOfWeekConstraint.createRange(lowerBound: lowerBound, upperBound: upperBound)
+
+        return self
+    }
+
+    public func whenDayOfMonthInRange(lowerBound: Int, upperBound: Int) throws -> Self {
+        try dayOfMonthConstraint.createRange(lowerBound: lowerBound, upperBound: upperBound)
+
+        return self
+    }
+
+    
+    public func whenMonthInRange(lowerBound: Int, upperBound: Int) throws -> Self {
+        try monthConstraint.createRange(lowerBound: lowerBound, upperBound: upperBound)
+
+        return self
+    }
+
+    public func whenQuarterInRange(lowerBound: Int, upperBound: Int) throws -> Self {
+        try quarterConstraint.createRange(lowerBound: lowerBound, upperBound: upperBound)
+
+        return self
+    }
+
+    public func whenYearInRange(lowerBound: Int, upperBound: Int) throws -> Self {
+        try yearConstraint.createRange(lowerBound: lowerBound, upperBound: upperBound)
+
+        return self
+    }
+
+    private func resolveLowestCadence() -> RecurrenceRuleTimeUnit {
+        if secondConstraint.isConstraintActive {
+            return .second
+        } else if minuteConstraint.isConstraintActive {
+            return .minute
+        } else if hourConstraint.isConstraintActive {
+            return .hour
+        } else if dayOfMonthConstraint.isConstraintActive {
+            return .dayOfMonth
+        } else if monthConstraint.isConstraintActive {
+            return .month
+        } else {
+            return .year
+        }
+    }
+
+    private func resolveCadenceLevel(_ ruleTimeUnit: RecurrenceRuleTimeUnit) -> Int {
+        switch ruleTimeUnit {
+        case .second:
+            return 0
+        case .minute:
+            return 1
+        case .hour:
+            return 2
+        case .dayOfMonth:
+            return 3
+        case .month:
+            return 4
+        default:
+            return 6
+        }
+    }
+
+    public func evaluate(date: Date) throws -> Bool {
+        return try evaluate(date: date).isValid
+    }
+
+    private func evaluate(date: Date) throws -> (isValid: Bool, ruleTimeUnitFailedOn: RecurrenceRuleTimeUnit?) {
```

Comment:
I'm fine with returning a tuple for now but it might be worth creating a separate object